File: week 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
int main(void)
{
    long card = get_long("Number: ");
    long n = card;
    int sum = 0;
    int length = 0;
    bool alt = false;
    while (n > 0)
    {
        int digit = n % 10;
        if (alt)
        {
            int prod = digit * 2;
            sum += prod / 10 + prod % 10;
        }
        else
        {
            sum += digit;
        }
        alt = !alt;
        n /= 10;
        length++;
    }
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    long first_two = card;
    while (first_two >= 100)
    {
        first_two /= 10;
    }
    long first_one = first_two / 10;
    if (length == 15 && (first_two == 34 || first_two == 37))
        printf("AMEX\n");
    else if (length == 16 && (first_two >= 51 && first_two <= 55))
        printf("MASTERCARD\n");
    else if ((length == 13 || length == 16) && first_one == 4)
        printf("VISA\n");
    else
        printf("INVALID\n");
}

File: week 6/sentimental-credit/credit.py
from cs50 import get_string
def main():
    card = get_string("Number: ")
    if not luhn_check(card):
        print("INVALID")
        return
    if (len(card) == 15) and (card.startswith("34") or card.startswith("37")):
        print("AMEX")
    elif (len(card) == 16) and (51 <= int(card[:2]) <= 55):
        print("MASTERCARD")
    elif (len(card) in [13, 16]) and card.startswith("4"):
        print("VISA")
    else:
        print("INVALID")
def luhn_check(number):
    """Return True if number is valid per Luhn's algorithm, else False."""
    total = 0
    reverse_digits = number[::-1]
    for i, digit in enumerate(reverse_digits):
        n = int(digit)
        if i % 2 == 1:
            n *= 2
            if n > 9:
                n -= 9
        total += n
    return total % 10 == 0
if __name__ == "__main__":
    main()
